framework/src/play/mvc/ActionInvoker.java

```diff
@@ -147,11 +147,15 @@ public static void invoke(Http.Request request, Http.Response response) {
 
                 // Check the cache (only for GET or HEAD)
                 if ((request.method.equals("GET") || request.method.equals("HEAD")) && actionMethod.isAnnotationPresent(CacheFor.class)) {
-                    cacheKey = actionMethod.getAnnotation(CacheFor.class).id();
+                    CacheFor cacheFor = actionMethod.getAnnotation(CacheFor.class);;
+                    cacheKey = cacheFor.id();
                     if ("".equals(cacheKey)) {
-                        cacheKey = "urlcache:" + request.url + request.querystring;
+                        // Generate a cache key for this request
+                        cacheKey = cacheFor.generator().newInstance().generate(request);
```

Comment:
Can't it cause a bigger memory footprint because of creating a new instance of generator every time?

Reply:
There two options here:
1) We can maintain a cache of CacheKeyGenerator instances in the ActionInvoker.
2) Push the creation of generator to be application developer responsibility by added a get() method to the CacheKeyGenerator interface.

public interface CacheKeyGenerator {
	public String generate(Request request);
        public CacheKeyGenerator get();
}

Line 154 will can rewitten as:

cacheKey = cacheFor.generator().get().generate(request);

Reply:
DefaultCacheKeyGenerator can implement a singleton 

```Java

public class DefaultCacheKeyGenerator implements CacheKeyGenerator {
        private final static CacheKeyGenerator instance;
	@Override
	public String generate(Request request) {
		return "urlcache:" + request.url + request.querystring;
	}

       public CacheKeyGenerator get() {
               if(instance == null) {
                        instance = new DefaultCacheKeyGenerator();
               }
               return  instance;
       }
}
```

Reply:
I can make quick these changes if this is what is required to get this pull request out of the door?
I am also all ears for any other solutions.

Reply:
@tahseenarticle I am not really sure if we need to change anything. I just asked :)
Probably it would be a premature optimization. 
What I actually am waiting is @xael-fry review.